[CL-3540] Remove openid_connect fork

adessy opened the pull request:
I tested it locally with Bosa FAS, but it might be worth testing it with Clave Unica as well. What do you think?

# Testing

- [x] Bosa FAS (verification)
- [x] ClaveUnica (verification)
- [x] FranceConnect (verification)
- [x] FranceConnect (authentication)

# Changelog
## Technical
- Remove `openid_connect` fork from the `Gemfile`.

adessy committed:
[CL-3540] Refactor Omniauth initializer

- Move OpenIDConnect patch to the gem_extensions folder.
- Reorder things in the initializer.

adessy committed:
[CL-3540] Remove openid_connect fork

We no longer need to explicitly specify openid_connect as a dependency because it is already a dependency of omniauth_openid_connect.

I upgraded omniauth_openid_connect to the latest version, which uses the most recent version of openid_connect. The latest version of openid_connect was refactored to use Faraday as the HTTP client, which allows us to use a Faraday middleware to address the issue we had with JWT encoding (which is why we forked openid_connect in the first place).

As a side effect, we no longer require the HTTPClient patch since we are now using Faraday as the HTTP client.

adessy committed:
Merge branch 'master' into CL-3540/remove-openid-connect-fork

# Conflicts:
#	back/Gemfile.lock

adessy commented on back/config/initializers/omniauth.rb:
This is the new bit. The rest is just refactoring.

adessy commented on front/app/containers/Admin/reporting/components/ReportBuilder/Templates/ProjectTemplate.tsx:
```diff
@@ -49,7 +49,8 @@ const ProjectTemplate = ({ reportId, projectId }: Props) => {
   );
 
   const hasPhases =
-    project.data.attributes.process_type === 'continuous' && phases.data.length > 0;
+    project.data.attributes.process_type === 'continuous' &&
+    phases.data.length > 0;
```
Not sure where this is coming from... 🧐 Probably an artifact of the conflict resolution with the master branch.

alexander-cit commented on back/engines/commercial/id_franceconnect/config/initializers/httpclient_patch.rb:
```diff
@@ -1,20 +0,0 @@
-# https://github.com/nahi/httpclient/pull/386 is merged, we work around
-# this limitation by forcing the HTTPClient SSL store to use the default
-# system configuration.
-module HTTPClient::SSLConfigDefaultPaths
```
Why do you remove it? Did you manage to reproduce the original issue in the old version and fix it in the new version? https://github.com/CitizenLabDotCo/citizenlab/pull/1590

alexander-cit replied:
Ah, just noticed it

> As a side effect, we no longer require the HTTPClient patch since we are now using Faraday as the HTTP client.

alexander-cit replied:
But anyway it may be useful to test that the original issue is still reproduced in master, but not in this branch.

alexander-cit reviewed (commented):
Yes, I think it makes a lot of sense to test FranceConnect (verification + authentication) and ClaveUnica (verification). There are very detailed instructions here https://www.notion.so/citizenlab/SSO-set-up-locally-and-test-a176dd3e335f47c28f949e5962d32a26

And if you feel like you have time, you can also test ClaveUnica authentication in this branch https://github.com/CitizenLabDotCo/citizenlab/pull/4825. If not, it's fine, I'll test it anyway once again before merging.

What credentials did you use to test Bosa FAS locally? If not your private ones, is it possible to add them to the doc and LastPass?

alexander-cit reviewed (approved):
The code looks good but probably requires pretty extensive testing.

alexander-cit commented:
In the next few months, I want to write some e2e tests for it so we won't need to do it manually. It should be [doable with Cypress](https://www.cypress.io/blog/2022/04/25/cypress-9-6-0-easily-test-multi-domain-workflows-with-cy-origin/). If not, then we could use Playwright or sth else.

adessy committed:
[CL-3540] Revert part of 3fce235a78b3b22071002f3ed02baa13847f1d3c

In the commit 3fce235a78b3b22071002f3ed02baa13847f1d3c, we made the assumption that:
> we no longer require the HTTPClient patch since we are now using Faraday as the HTTP client.

However, while Faraday is used for rack-oauth2 queries, google-apis-core still relies on HTTPClient. Although the certificate problem does not appear to occur with google-apis-core, we have decided to keep the patch as a precautionary measure as long as some dependencies rely on httpclient.

Refer to commit 352fc2c1cf54b3b4ce4dc04e4041306278aa69a8 for more context on why HTTPClient is patched.

adessy commented:
> What credentials did you use to test Bosa FAS locally? If not your private ones, is it possible to add them to the doc and LastPass?

The process will not be easy to reproduce. I basically used the credentials of an active production platform and spoofed the platform locally by modifying the `/etc/hosts` file and running an nginx server to proxy the requests to localhost:3000. Finally, I used my own identity to test the verification. Perks of being a Belgian citizen :)
